### PR TITLE
Cockpit: limit number of shown decimal places when formatting numbers

### PR DESCRIPTION
**Pull request by dperpeet**
See testcase
555.98765432, "556.0 bps"
is this what we want?

For a "clean" solution we'd have to check if the rounded result is an integer (or even === factor) and change things accordingly.

I believe this output is ok, since it indicates that higher precision than integer is at play and that the result may have been affected by rounding.

### CODE REVIEW

**Comment by andreasn** [comment]
I think this makes sense, yes.

**Comment by dperpeet** [comment]
result of irc discussion: ".0" at the end of a number is redundant.

**Comment by stefwalter** [comment]
> result of irc discussion: ".0" at the end of a number is redundant.

It is only redundant when the result is exactly X of the factor. If it has been rounded to X, then we should use X.0

An examples:

 * 1024 bytes -> 1 KB
 * 1025 bytes -> 1.0 KB

File: pkg/base1/cockpit.js [hunk 1]
```diff
@@ -2163,33 +2161,34 @@ function full_scope(cockpit, $, po) {
...
+        var ret;
 
         /* TODO: Make the decimal separator translatable */
-        if (number === 0 || !divided)
-            ret = [number.toString(), suffix];
-        else
+        if (suffix)
             ret = [number.toFixed(1), suffix];
+        else
+            ret = [number.toFixed(1)];
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: This changes the API. This API is meant to return two values unless ```!separate```. What's the logic behind this change. This change is also not documented.

Comment (dperpeet) [inline pkg/base1/cockpit.js]: > This changes the API. This API is meant to return two values unless !separate. What's the logic behind this change. This change is also not documented.

This is not an API change, just moving inside the function. [See current master](https://github.com/cockpit-project/cockpit/blob/293d8ce1e6cd21970dfb73ea5a8667d5fc78b39d/pkg/base1/cockpit.js#L2175-L2180)

**Comment by dperpeet** [comment]
> It is only redundant when the result is exactly X of the factor. If it has been rounded to X, then we should use X.0

I was of your opinion, but changed my mind after discussing with Andreas and Marius. After all, reading the trailing .0 seems redundant. Marius pointed out that the distinction we make may not be clear to the user anyway.

@andreasn How should we proceed on this?

File: pkg/base1/cockpit.js (hunk 1 above)
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Aha, makes sense.

**Comment by dperpeet** [comment]
I think the best choice is to show natural numbers as such and everything else in fixed point notation, even with a trailing zero. For this it doesn't matter if the natural number resulted from division or not.

Example: If the result of the division is a natural number - at least within double precision - we should consider it as such (e.g. 2048 / 1024 is 2, not 2.0).

**Comment by dperpeet** [comment]
As an alternative we could say that we drop the decimal place completely for values >= 10.
This is equivalent to saying we want at least 2 significant digits: 5.5, 9.8, 11

**Comment by andreasn** [comment]
My biggest concern right now is the numbers on the left of the graphs in #2020
In general elsewhere, like on the top right of the same graphs I would be more OK with seeing a 3.0 or so.

**Comment by stefwalter** [comment]
> My biggest concern right now is the numbers on the left of the graphs in #2020

First of all, I if it's exactly N units, we should display N and not N.0 ... The latter should be displayed when we discard information during rounding.

And as noted in #2020, I think that should be fixed in the graph itself to display Y axis grid lines intelligently at N units, similar to how we display lines intelligently on the X axis intelligently (on minutes).

**Comment by dperpeet** [comment]
> First of all, I if it's exactly N units, we should display N and not N.0 ... The latter should be displayed when we discard information during rounding.

The current implementation ignores information lost due to js internal number representation. I believe this to be acceptable and even necessary when dealing with large integers - since js only knows 64 bit fp.

**Comment by mvollmer** [comment]
> As an alternative we could say that we drop the decimal place completely for values >= 10.
This is equivalent to saying we want at least 2 significant digits: 5.5, 9.8, 11

JavaScript has `X.toPrecision(2)`, which produces 5.5 and 11, but it might also produce 0.30 instead of 0.3, i.e., it doesn't count a leading "0.".

**Comment by dperpeet** [comment]
> JavaScript has X.toPrecision(2), which produces 5.5 and 11, but it might also produce 0.30 instead of 0.3, i.e., it doesn't count a leading "0.".

even worse. toPrecision can produce scientific notation for numbers larger than the precision you pass (e.g. 937 could become 9.4e+2)

**Comment by dperpeet** [comment]
@andreasn @stefwalter how should we proceed?
I would be ok leaving this as it is in the general case.
Options I could add:
- flag to remove trailing .0 if present
- [change default to / add flag to] show decimal place only for values <10

**Comment by stefwalter** [comment]
I'll work on this. We should detect exact multiples of units, and drop the .0 in that case.

**Comment by dperpeet** [comment]
@stefwalter exact multiples will be detected and are without .0 - all natural numbers are converted using standard number-to-string, not fixed notation:
https://github.com/cockpit-project/cockpit/pull/2182/files#diff-34cbdb779d20a4a75591fcba50bd25d4R2182

**Comment by stefwalter** [comment]
You're right. I added more tests, and merged.

**Pull request closed without merging**